mono/metadata/sgen-thread-pool.c

```diff
@@ -39,6 +39,9 @@ static SgenThreadPoolThreadInitFunc thread_init_func;
 static SgenThreadPoolIdleJobFunc idle_job_func;
 static SgenThreadPoolContinueIdleJobFunc continue_idle_job_func;
 
+static int threadpool_shutdown;
+static int thread_finished;
```

Comment:
A sufficiently smart whole program optimizer could figure this out, which would break the code.  If nothing else, `volatile` reminds us that these are accessed from different threads at the same time, and gives us peace of mind.  Concurrent programming is hard enough as it is, without depending on the goodwill of the compiler.